### PR TITLE
JDK-8295236: Update JavaDoc in javafx.geometry.Point3D

### PR DESCRIPTION
**Pull request by douglasheldoracle**
The JavaDoc for equals had a copy/paste error. I normalized the text based on the JavaDoc for method java.awt.Point#equals. ~~I also changed formatting in the method signatures of equals(), hashCode() and toString().~~

For good measure, some kind of copy/paste detection should probably be added to the many automated checks. For the entire OpenJDK project.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8295236](https://bugs.openjdk.org/browse/JDK-8295236): Update JavaDoc in javafx.geometry.Point3D

### Reviewers
 * [Nir Lisker](https://openjdk.org/census#nlisker) (@nlisker - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx pull/913/head:pull/913` \
`$ git checkout pull/913`

Update a local copy of the PR: \
`$ git checkout pull/913` \
`$ git pull https://git.openjdk.org/jfx pull/913/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 913`

View PR using the GUI difftool: \
`$ git pr show -t 913`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/913.diff">https://git.openjdk.org/jfx/pull/913.diff</a>

</details>

### CODE REVIEW

**Comment by kevinrushforth** [comment]
This is now available in JBS as [JDK-8295236](https://bugs.openjdk.org/browse/JDK-8295236). Once you update the title of this PR (and fix your whitespace errors), it will be made `rfr`.

**Comment by kevinrushforth** [comment]
Looks fine to me, but I'll let @arapte review and sponsor this.

File: modules/javafx.graphics/src/main/java/javafx/geometry/Point3D.java
Comment (nlisker) [inline modules/javafx.graphics/src/main/java/javafx/geometry/Point3D.java:413]: Since this object is known, I would start with "Determines if this point is equals to a given object".

File: modules/javafx.graphics/src/main/java/javafx/geometry/Point3D.java
Comment (nlisker) [inline modules/javafx.graphics/src/main/java/javafx/geometry/Point3D.java:414]: I would even write "i.f.f" because it's bidirectional.

Also, `x`, `y`, and `z` should be in `{@code}`.

File: modules/javafx.graphics/src/main/java/javafx/geometry/Point3D.java
Comment (nlisker) [inline modules/javafx.graphics/src/main/java/javafx/geometry/Point3D.java:415]: We tend to put a new line before the tags to separate them from the description, not sure if it's enforced.

File: modules/javafx.graphics/src/main/java/javafx/geometry/Point3D.java
Comment (nlisker) [inline modules/javafx.graphics/src/main/java/javafx/geometry/Point3D.java:417]: `true`, `Point3D` and `false` should be in `{@code}`.

**Review by nlisker — CHANGES_REQUESTED** [review]

File: modules/javafx.graphics/src/main/java/javafx/geometry/Point3D.java
Comment (kevinrushforth) [reply modules/javafx.graphics/src/main/java/javafx/geometry/Point3D.java:414]: I think this one is OK either as "if" or "if and only if" (if you do change it, I recommend spelling it out, although "iff" with no punctuation, would be acceptable). I agree with the request to use code style.

File: modules/javafx.graphics/src/main/java/javafx/geometry/Point3D.java
Comment (kevinrushforth) [reply modules/javafx.graphics/src/main/java/javafx/geometry/Point3D.java:415]: Not strictly enforced, but adding a blank line does aid readability (of the source code...the generated docs don't care).

**Review by kevinrushforth — COMMENTED** [review]
A couple comments.

File: modules/javafx.graphics/src/main/java/javafx/geometry/Point3D.java
Comment (douglasheldoracle) [reply modules/javafx.graphics/src/main/java/javafx/geometry/Point3D.java:413]: Agree. I propose the very similar wording "Determines whether this Point3D is equal to a given object."

File: modules/javafx.graphics/src/main/java/javafx/geometry/Point3D.java
Comment (douglasheldoracle) [reply modules/javafx.graphics/src/main/java/javafx/geometry/Point3D.java:414]: I disagree with both suggestions. 

1. I think the extra verbosity (while factually correct) adds nothing, for such a straightforward concept as equality of points in 3 space.
2. x, y and z are private and are not exposed in the JavaDoc. Further, the exposed methods getX() etc. that I **could** link to, are not actually what are evaluated in the method body.

I'd propose alternately that we keep "x, y and z" but call them "coordinates" and not "properties". This agrees with other parts of the JavaDoc such as the constructor, and getX() etc. method documentation.

File: modules/javafx.graphics/src/main/java/javafx/geometry/Point3D.java
Comment (douglasheldoracle) [reply modules/javafx.graphics/src/main/java/javafx/geometry/Point3D.java:415]: If I were to add empty lines before JavaDoc tags such as @param etc, then for consistency I think I should do it for the whole class, where they are not present.

Note, on line 32: **// PENDING_DOC_REVIEW of this whole class**
Maybe we should first decide whether we are scrubbing all doc for the whole class, or making a minimal change to the obviously incorrect information. My feeling is now tending toward the latter.

It could very well be that I misunderstood the comment by nlisker.

File: modules/javafx.graphics/src/main/java/javafx/geometry/Point3D.java
Comment (douglasheldoracle) [reply modules/javafx.graphics/src/main/java/javafx/geometry/Point3D.java:417]: ...correction, found in toString() as well.

File: modules/javafx.graphics/src/main/java/javafx/geometry/Point3D.java
Comment (douglasheldoracle) [reply modules/javafx.graphics/src/main/java/javafx/geometry/Point3D.java:414]: Correction: getX() etc. are used in the method. I will put the methods into code tags.

**Comment by douglasheldoracle** [comment]
I've reverted some of my changes which fell outside of the egregious documentation error.

If we are to correct further doc style, then that might be addressed as a whole-file review of the doc. (per the comment on line 32)

File: modules/javafx.graphics/src/main/java/javafx/geometry/Point3D.java
Comment (nlisker) [reply modules/javafx.graphics/src/main/java/javafx/geometry/Point3D.java:418]: I think that we should explain what makes 2 objects equal. Even if we don't explicitly name the methods used for comparison, we could say "2 points are equals if their coordinates are equal".

By the way, I have https://bugs.openjdk.org/browse/JDK-8226930 assigned to go over some dubious equals/hashcode implementations, in case you want to delegate the task.

Comment (arapte) [reply modules/javafx.graphics/src/main/java/javafx/geometry/Point3D.java:418]: Agreed, mentioning equality criteria sounds good to me too. How does this look ?

```
    /**
     * Indicates whether some other object is "equal to" this one.
     * Two instances of Point3D are equal if the return values of their
     * {@code getX}, {@code getY}, and {@code getZ} methods are equal.
     *
     * @param obj the reference object with which to compare
     * @return true if this Point3D is the same as the obj argument; false otherwise
     */

```

Comment (nlisker) [reply modules/javafx.graphics/src/main/java/javafx/geometry/Point3D.java:418]: This is the current proposal by Douglas and align with `Object::equals` which it is overriding. I'm fine with this wording. Just need to add the `@code` tags.

Comment (douglasheldoracle) [reply modules/javafx.graphics/src/main/java/javafx/geometry/Point3D.java]: I have added a commit to match the proposed block by arapte.

**Comment by douglasheldoracle** [comment]
Tagged @nlisker to validate the code tags are included appropriately.

File: modules/javafx.graphics/src/main/java/javafx/geometry/Point3D.java
Comment (nlisker) [inline modules/javafx.graphics/src/main/java/javafx/geometry/Point3D.java:414]: `Point3D` also needs `@code`.

File: modules/javafx.graphics/src/main/java/javafx/geometry/Point3D.java
Comment (nlisker) [inline modules/javafx.graphics/src/main/java/javafx/geometry/Point3D.java:418]: `Point3D`, `true`, `false` and `obj` need `@code`.

**Comment by douglasheldoracle** [comment]
I think this equals() method JavaDoc should be compliant now.

I must apologise. I thought the code tags produced hyperlinks. Now I can see they only apply a fixed width type font. Of course, the formatting tags need to apply in all of the places discussed.

File: modules/javafx.graphics/src/main/java/javafx/geometry/Point3D.java
Comment (nlisker) [inline modules/javafx.graphics/src/main/java/javafx/geometry/Point3D.java:417]: Don't use `@code` for the parameter declaration, they are treated automatically by the javadoc tool.

Comment (douglasheldoracle) [reply modules/javafx.graphics/src/main/java/javafx/geometry/Point3D.java:417]: Don't use it for `obj` you mean?

**Review by nlisker — APPROVED** [review]

**Comment by nlisker** [comment]
> I thought the code tags produced hyperlinks.

Those would be the `@link` tags, or `@linkplain`.

**Comment by douglasheldoracle** [comment]
I think there's something else of quality worth discussing here. Is `==` appropriate given the values under comparison are `double`?

**Comment by nlisker** [comment]
> I think there's something else of quality worth discussing here. Is `==` appropriate given the values under comparison are `double`?

It's not. I have mentioned this exact case in https://bugs.openjdk.org/browse/JDK-8226930. I can assign it to you (or self-assign it yourself) if you want to work on it. If not, I will get to it at some point (no pun intended). This issue will require to go over the classes that override equals and see if they do it properly, including `hashcode`. Sometimes the `equals` implementation is "good enough", so there's not need to rewrite it in all the classes, but especially for public APIs, users should expect correct implementations.

**Comment by douglasheldoracle** [comment]
I wouldn't want to take on this fix, as I'm not well educated in floating point types. I'm glad it is an identified issue.

--
Douglas Held
Senior Principal Security Consultant
NetSuite Product Security Team
***@***.***

Note: Sent from a phone/microphone which may have introduced errors I didn’t catch

On 19 Oct 2022, at 13:52, nlisker ***@***.***> wrote:

﻿

I think there's something else of quality worth discussing here. Is == appropriate given the values under comparison are double?

It's not. I have mentioned this exact case in https://bugs.openjdk.org/browse/JDK-8226930. I can assign it to you (or self-assign it yourself) if you want to work on it. If not, I will get to it at some point (no pun intended). This issue will require to go over the classes that override equals and see if they do it properly, including hashcode. Sometimes the equals implementation is "good enough", so there's not need to rewrite it in all the classes, but especially for public APIs, users should expect correct implementations.

**Comment by kevinrushforth** [comment]
> It's not. I have mentioned this exact case in https://bugs.openjdk.org/browse/JDK-8226930. I can assign it to you (or self-assign it yourself) if you want to work on it. If not, I will get to it at some point (no pun intended). This issue will require to go over the classes that override equals and see if they do it properly, including `hashcode`. Sometimes the `equals` implementation is "good enough", so there's no need to rewrite it in all the classes, but especially for public APIs, users should expect correct implementations.

And it's not just a doc issue: Point3D (at least) violates the contract of `equals` and `hashCode`, since two objects that are considered `equals` can produce different hash codes (e.g., the case of two points, one of which has a value of `+0.0` for one of the coords, and the other of which has `-0.0` for the same coord). Depending on how we change it, it might need a CSR.

So I'd prefer that either you keep the bug or @arapte take it.

**Comment by nlisker** [comment]
I can keep it.

**Pull request closed without merging**
